src/OrchardCore/OrchardCore.Infrastructure.Abstractions/Security/Permissions/DefaultPermissionService.cs

```diff
@@ -0,0 +1,58 @@
+namespace OrchardCore.Security.Permissions;
+
+public sealed class DefaultPermissionService : IPermissionService
+{
+    private readonly IEnumerable<IPermissionProvider> _permissionProviders;
+
+    // Cached per request.
+    private Dictionary<string, Permission> _permissions;
+
+    public DefaultPermissionService(IEnumerable<IPermissionProvider> permissionProviders)
+    {
+        _permissionProviders = permissionProviders;
+    }
+
+    public async ValueTask<Permission> FindByNameAsync(string name)
+    {
+        ArgumentException.ThrowIfNullOrEmpty(name);
+
+        if (_permissions is null)
```

Comment:
What about using the dictionary if it's initialized, and just iterating the providers if not? This way the first match will return the permission.

Reply:
What if we don't load the list first, that will cause enumeration for each name lookup which will defeat the purpose.